### ISSUE TITLE
CUDA OOM issue when running tests in CI

### ISSUE THREAD

**Issue by huydhn** [pytorch/pytorch#85939]
### 🐛 Describe the bug

There have been reports of OOM issue when running CUDA tests in CI. For example,

* https://github.com/pytorch/pytorch/actions/runs/3146997338/jobs/5116296083

As the first step, we need to gather more examples to see if the same test fails always or if it's a more widespread issue because Linux CUDA test_ops are now run in parallel (https://github.com/pytorch/pytorch/pull/85528, in 3 processes)

### Versions

CUDA 11.6 (and may be 10.2 too)

cc @ezyang @gchanan @zou3519 @seemethere @malfet @pytorch/pytorch-dev-infra @clee2000

**Comment by kwen2501** [comment]
I hit an OOM job as well:
https://github.com/pytorch/pytorch/actions/runs/3154594322/jobs/5132619975

**Comment by huydhn** [comment]
I think that we can also get the data points by querying the logs from Rockset.  Let me verify this.

```
2022-09-29T21:50:44.3353115Z =================================== FAILURES ===================================
2022-09-29T21:50:44.3353336Z _______ TestCommonCUDA.test_python_ref__refs_float_power_cuda_complex128 _______
2022-09-29T21:50:44.3353483Z Traceback (most recent call last):
2022-09-29T21:50:44.3353700Z   File "/var/lib/jenkins/workspace/test/test_ops.py", line 343, in test_python_ref
2022-09-29T21:50:44.3353918Z     self._ref_test_helper(lambda: TorchRefsMode(strict=True), device, dtype, op)
2022-09-29T21:50:44.3354117Z   File "/var/lib/jenkins/workspace/test/test_ops.py", line 219, in _ref_test_helper
2022-09-29T21:50:44.3354328Z     for sample in op.reference_inputs(device, dtype, requires_grad=False):
2022-09-29T21:50:44.3354765Z   File "/opt/conda/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 43, in generator_context
2022-09-29T21:50:44.3354904Z     response = gen.send(None)
2022-09-29T21:50:44.3355331Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 1445, in reference_inputs_elementwise_binary
2022-09-29T21:50:44.3355458Z     yield from gen()
2022-09-29T21:50:44.3355893Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 1391, in _reference_inputs_elementwise_binary
2022-09-29T21:50:44.3356113Z     yield from op.sample_inputs_func(op, device, dtype, requires_grad, **kwargs)
2022-09-29T21:50:44.3356444Z   File "/opt/conda/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 43, in generator_context
2022-09-29T21:50:44.3356579Z     response = gen.send(None)
2022-09-29T21:50:44.3356995Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 1903, in sample_inputs_elementwise_binary
2022-09-29T21:50:44.3357175Z     lhs = make_arg(shape_lhs, **op.lhs_make_tensor_kwargs)
2022-09-29T21:50:44.3357524Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_creation.py", line 138, in make_tensor
2022-09-29T21:50:44.3357733Z     real_rand_val = torch.rand(shape, device=device, dtype=float_dtype)
2022-09-29T21:50:44.3358331Z torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.00 MiB (GPU 0; 7.44 GiB total capacity; 351.24 MiB already allocated; 192.00 KiB free; 1.64 GiB allowed; 888.00 MiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by clee2000** [comment]
another data point https://github.com/pytorch/pytorch/actions/runs/3154840440/jobs/5133121944

**Comment by ptrblck** [comment]
The error message:
```
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.00 MiB (GPU 0; 7.44 GiB total capacity; 351.24 MiB already allocated; 192.00 KiB free; 1.64 GiB allowed; 888.00 MiB reserved in total by PyTorch) 
```
indicates that `torch.cuda.set_per_process_memory_fraction` might have been used as only `1.64 GiB` are "allowed", see:
```python
x = torch.randn(10000000000, device='cuda')
# RuntimeError: CUDA out of memory. Tried to allocate 37.25 GiB (GPU 0; 23.68 GiB total capacity; 192.29 MiB already allocated; 15.77 GiB free; 5.05 GiB reserved in total by PyTorch)

torch.cuda.set_per_process_memory_fraction(0.1)
x = torch.randn(10000000000, device='cuda')
# RuntimeError: CUDA out of memory. Tried to allocate 37.25 GiB (GPU 0; 23.68 GiB total capacity; 192.29 MiB already allocated; 15.79 GiB free; 2.37 GiB allowed; 5.05 GiB reserved in total by PyTorch)
```

I don't see its usage in `test_ops.py` (`test_cuda.py` tests the function itself), but these lines of code in [`torch/testing/_internal/cmmon_utils.py`](https://github.com/pytorch/pytorch/blob/dab1c7c379d9597d8424aa55da888cfabe9ade3b/torch/testing/_internal/common_utils.py#L910-L913) might interact with this test:
```python
if TEST_CUDA and 'NUM_PARALLEL_PROCS' in os.environ:
    num_procs = int(os.getenv("NUM_PARALLEL_PROCS", "3"))
    # other libraries take up about 11% of space per process
    torch.cuda.set_per_process_memory_fraction(round(1 / num_procs - .11, 2))
```
which reduce the max. memory based on parallel test runs and were introduced a week ago in: https://github.com/pytorch/pytorch/pull/85528

**Comment by ezyang** [comment]
Still happening on master https://github.com/pytorch/pytorch/actions/runs/3156343744/jobs/5136168288

**Comment by huydhn** [comment]
@clee2000 Do you think we should reduce the number of parallel processes from 3 to 2 here to address the issue?  At the moment, we don't have the exact profile of GPU memory usage yet.  It could be that the memory usage increases between https://github.com/pytorch/pytorch/pull/85528 and today, thus tipping thing over the top

**Comment by huydhn** [comment]
> ```python
> if TEST_CUDA and 'NUM_PARALLEL_PROCS' in os.environ:
>     num_procs = int(os.getenv("NUM_PARALLEL_PROCS", "3"))
>     # other libraries take up about 11% of space per process
>     torch.cuda.set_per_process_memory_fraction(round(1 / num_procs - .11, 2))
> ```

@ptrblck As an immediate mitigation, we have reduced the number of process from 3 to 2. Do you have any recommendations for us on how to efficiently utilize GPU memory when things are run in parallel? IMO, having a fixed thresholds like this sounds a bit brittle

**Comment by huydhn** [comment]
Paste a note here from our internal discussion about a potential memory leak issue in https://github.com/pytorch/pytorch/commit/b04b2fa9aa52cacbdc9aaaf477d55b0af845ce81#diff-893b1eea27352f336f4cd832919e48d721e4e90186e63400b8596db6b82e7450R3155

**Comment by eqy** [comment]
> @clee2000 Do you think we should reduce the number of parallel processes from 3 to 2 here to address the issue? At the moment, we don't have the exact profile of GPU memory usage yet without spending efforts digging around. It could be that the memory usage increases between #85528 and today, thus tipping thing over the top

Yes, I did notice some flaky failures while testing this PR in CI, but it was tricky as it didn't seem to happen on every instance (perhaps some GPU runners have more memory or the issue isn't deterministic due to parallelism).

As for skipping the leak check, thanks for the catch---it was leftover and outdated from when the PR did not clear caches before and after each test. I have opened a new PR for this as the leak check should now pass:
https://github.com/pytorch/pytorch/pull/85988

**Comment by huydhn** [comment]
As a mitigation, we have done:

* Reduce the number of parallel process from 3 to 2 https://github.com/pytorch/pytorch/pull/85985
* Revert the recent change https://github.com/pytorch/pytorch/pull/85447#issuecomment-1264014341 that might potentially cause CUDA memory leak

Please do a `rebase -m` to rebase to master to get these changes and let's know how it goes

**Comment by ptrblck** [comment]
> @ptrblck As an immediate mitigation, we have reduced the number of process from 3 to 2. Do you have any recommendations for us on how to efficiently utilize GPU memory when things are run in parallel? IMO, having a fixed thresholds like this sounds a bit brittle

Yes, I agree that a 3-way split might be too optimistic. A better approach could be to:
* split the GPU memory depending on its general available memory (e.g. splitting a GPU with only 7.5GB might not allow you to run a lot of tests) and a threshold could be used (e.g. >= 20GB cards)
* try to profile the unit tests to see the overall memory usage (would anyone have an idea how we could do it?)
* exclude performance tests and guarantee an exclusive run on the GPU to avoid noisy reports

Thanks for reducing the split to 2 runs. As we are close to the 1.13.0 cut, we might also consider reverting the parallelization for now and follow up after the branch cut in case tests are still failing. What do you think?

**Comment by eqy** [comment]
@huydhn @ngimel
Was there a root-cause/evidence for the memory leak issue? I see that a `torchbenchmark` issue was raised, but I believe it would be fixable by clearing the cuBLAS workspaces before and after the model run: https://github.com/pytorch/pytorch/pull/85447#issuecomment-1264038375

Is the current understanding that there is an actual memory leak (growing over time) or that the additional memory usage tracked by `CUDACachingAllocator` rather than being managed by cuBLAS is stepping over the allowed memory limit?

**Comment by huydhn** [comment]
> Thanks for reducing the split to 2 runs. As we are close to the 1.13.0 cut, we might also consider reverting the parallelization for now and follow up after the branch cut in case tests are still failing. What do you think?

I'll keep this in mind to monitor trunk very closely in the next few days, we will revert the change if the issue persists.

**Comment by weiwangmeta** [comment]
Do we have more sightings of OOM? Trying to assess whether (after reverting https://github.com/pytorch/pytorch/pull/85447) we still see such issues or not

**Comment by huydhn** [comment]
> Do we have more sightings of OOM? Trying to assess whether (after reverting #85447) we still see such issues or not

I haven't seen OOM issue in the past few days

**Comment by huydhn** [comment]
I'm closing this issue now as this issue has been resolved.
